x-pack/plugins/fleet/public/applications/integrations/app.tsx

```diff
@@ -0,0 +1,233 @@
+import type { AppMountParameters } from 'kibana/public';
+import { EuiCode, EuiEmptyPrompt, EuiErrorBoundary, EuiPanel } from '@elastic/eui';
+import type { History } from 'history';
+import { createHashHistory } from 'history';
+import { Router, Redirect, Route, Switch } from 'react-router-dom';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { i18n } from '@kbn/i18n';
+import styled from 'styled-components';
+import useObservable from 'react-use/lib/useObservable';
+
+import {
+  ConfigContext,
+  FleetStatusProvider,
+  KibanaVersionContext,
+  sendGetPermissionsCheck,
+  sendSetup,
+} from '../../hooks';
+
+import type { FleetConfigType, FleetStartServices } from '../../plugin';
+
+import { KibanaContextProvider } from '../../../../../../src/plugins/kibana_react/public';
+import { EuiThemeProvider } from '../../../../../../src/plugins/kibana_react/common';
+
+import { INTEGRATIONS_ROUTING_PATHS } from './constants';
+
+import { Error, Loading } from './components';
+
+import type { UIExtensionsStorage } from './types';
+
+import { EPMApp } from './sections/epm';
+import { DefaultLayout, WithoutHeaderLayout } from './layouts';
+import { PackageInstallProvider } from './hooks';
+import { useBreadcrumbs, IntraAppStateProvider, UIExtensionsContext } from './hooks';
+
+const ErrorLayout = ({ children }: { children: JSX.Element }) => (
+  <EuiErrorBoundary>
+    <DefaultLayout>
+      <WithoutHeaderLayout>{children}</WithoutHeaderLayout>
+    </DefaultLayout>
+  </EuiErrorBoundary>
+);
+
+const Panel = styled(EuiPanel)`
+  max-width: 500px;
+  margin-right: auto;
+  margin-left: auto;
+`;
+
+export const WithPermissionsAndSetup: React.FC = memo(({ children }) => {
+  useBreadcrumbs('integrations');
+
+  const [isPermissionsLoading, setIsPermissionsLoading] = useState<boolean>(false);
+  const [permissionsError, setPermissionsError] = useState<string>();
+  const [isInitialized, setIsInitialized] = useState(false);
+  const [initializationError, setInitializationError] = useState<Error | null>(null);
+
+  useEffect(() => {
+    (async () => {
+      setIsPermissionsLoading(false);
+      setPermissionsError(undefined);
+      setIsInitialized(false);
+      setInitializationError(null);
+      try {
+        setIsPermissionsLoading(true);
+        const permissionsResponse = await sendGetPermissionsCheck();
+        setIsPermissionsLoading(false);
+        if (permissionsResponse.data?.success) {
+          try {
+            const setupResponse = await sendSetup();
+            if (setupResponse.error) {
+              setInitializationError(setupResponse.error);
+            }
+          } catch (err) {
+            setInitializationError(err);
+          }
+          setIsInitialized(true);
+        } else {
+          setPermissionsError(permissionsResponse.data?.error || 'REQUEST_ERROR');
+        }
+      } catch (err) {
+        setPermissionsError('REQUEST_ERROR');
+      }
```

Comment:
Thank you for the snippet, I definitely agree with this change! This logic could stand to be "unwrapped" a bit. I do think it might be best to push any refactors out of this initial PR, and to make a note of them for follow-ups. I worry that because of the large scope of changes here, a regression caused by an errant refactor might go unnoticed.

I've made a note of this, though, and will address it in the next set of changes.